### PR TITLE
sds: handle bogus file locations in destination rules

### PR DESCRIPTION
**Pull request by ramaraochavali**
Fixes https://github.com/istio/istio/issues/23822

### CODE REVIEW

File: security/pkg/nodeagent/cache/secretcache.go [hunk 1]
```diff
@@ -791,35 +791,38 @@ func (sc *SecretCache) generateFileSecret(connKey ConnKey, token string) (bool,
...
+		if sitem, err = sc.generateKeyCertFromExistingFiles(sc.existingCertChainFile, sc.existingKeyFile, token, connKey); err != nil {
+			// Adding cert is sufficient here as key can't change without changing the cert.
+			sc.addFileWatcher(sc.existingCertChainFile, token, connKey)
+		}
 	default:
 		// Check if the resource name refers to a file mounted certificate.
 		// Currently used in destination rules and server certs (via metadata).
 		// Based on the resource name, we need to read the secret from a file encoded in the resource name.
 		cfg, ok := pilotmodel.SdsCertificateConfigFromResourceName(connKey.ResourceName)
+		sdsFromFile = ok
```
Comment (ramaraochavali) [inline security/pkg/nodeagent/cache/secretcache.go:815]: @howardjohn This is the actual fix. We are setting sdsFromFile to true only if file exists in the below switch case. So after this method, sdsFromFile is false if file does not exist and getting in to other generation path.

**Comment by ramaraochavali** [comment]
@howardjohn  PTAL and see if you can test the fix with your setup. I have added a test to verify the case though

**Comment by ramaraochavali** [comment]
@howardjohn Sorry, Can you try again? This should fix.

**Comment by howardjohn** [comment]
ok looks like its working now. 

Previous behavior was file does not exist => LDS NACK => pod does not start up
New behavior is file does not exist => LDS warming => after 15s timeout, pod starts up

New behavior seems like an improvement.

One thing I noticed is we pretty heavily retry:

```
2020-05-19T15:16:29.675099Z     info    sds     resource:file-cert:deny~deny new connection
2020-05-19T15:16:29.675377Z     error   cache   resource:file-cert:deny~deny failed to generate secret for proxy from file: open deny: no such file or directory
2020-05-19T15:16:29.675440Z     error   sds     resource:file-cert:deny~deny Close connection. Failed to get secret for proxy "sidecar~10.244.0.9~alpine-554487f8f8-57vtc.default~default.svc.cluster.local" from secret cache: open deny: no such file or directory
2020-05-19T15:16:29.675575Z     info    sds     resource:file-cert:deny~deny connection is terminated: rpc error: code = Canceled desc = context canceled
2020-05-19T15:16:29.675895Z     warning envoy config    [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:92] StreamSecrets gRPC config stream closed: 2, open deny: no such file or directory
2020-05-19T15:16:33.164136Z     info    sds     resource:file-cert:deny~deny new connection
2020-05-19T15:16:33.164418Z     error   cache   resource:file-cert:deny~deny failed to generate secret for proxy from file: open deny: no such file or directory
2020-05-19T15:16:33.164470Z     error   sds     resource:file-cert:deny~deny Close connection. Failed to get secret for proxy "sidecar~10.244.0.9~alpine-554487f8f8-57vtc.default~default.svc.cluster.local" from secret cache: open deny: no such file or directory
2020-05-19T15:16:33.164582Z     info    sds     resource:file-cert:deny~deny connection is terminated: rpc error: code = Canceled desc = context canceled
2020-05-19T15:16:33.164932Z     warning envoy config    [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:92] StreamSecrets gRPC config stream closed: 2, open deny: no such file or directory
2020-05-19T15:16:36.095500Z     info    sds     resource:file-root:deny new connection
2020-05-19T15:16:36.095766Z     error   cache   resource:file-root:deny failed to generate secret for proxy from file: open deny: no such file or directory
2020-05-19T15:16:36.095833Z     error   sds     resource:file-root:deny Close connection. Failed to get secret for proxy "sidecar~10.244.0.9~alpine-554487f8f8-57vtc.default~default.svc.cluster.local" from secret cache: open deny: no such file or directory
2020-05-19T15:16:36.096076Z     info    sds     resource:file-root:deny connection is terminated: rpc error: code = Canceled desc = context canceled
2020-05-19T15:16:36.096395Z     warning envoy config    [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:92] StreamSecrets gRPC config stream closed: 2, open deny: no such file or directory
2020-05-19T15:16:38.835284Z     info    sds     resource:file-cert:deny~deny new connection
2020-05-19T15:16:38.835570Z     error   cache   resource:file-cert:deny~deny failed to generate secret for proxy from file: open deny: no such file or directory
2020-05-19T15:16:38.835626Z     error   sds     resource:file-cert:deny~deny Close connection. Failed to get secret for proxy "sidecar~10.244.0.9~alpine-554487f8f8-57vtc.default~default.svc.cluster.local" from secret cache: open deny: no such file or directory
2020-05-19T15:16:38.835723Z     info    sds     resource:file-cert:deny~deny connection is terminated: rpc error: code = Canceled desc = context canceled
2020-05-19T15:16:38.836047Z     warning envoy config    [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:92] StreamSecrets gRPC config stream closed: 2, open deny: no such file or directory
2020-05-19T15:16:45.211409Z     info    sds     resource:file-cert:deny~deny new connection
2020-05-19T15:16:45.211768Z     error   cache   resource:file-cert:deny~deny failed to generate secret for proxy from file: open deny: no such file or directory
2020-05-19T15:16:45.211824Z     error   sds     resource:file-cert:deny~deny Close connection. Failed to get secret for proxy "sidecar~10.244.0.9~alpine-554487f8f8-57vtc.default~default.svc.cluster.local" from secret cache: open deny: no such file or directory
2020-05-19T15:16:45.211951Z     info    sds     resource:file-cert:deny~deny connection is terminated: rpc error: code = Canceled desc = context canceled
2020-05-19T15:16:45.212368Z     warning envoy config    [bazel-out/k8-opt/bin/external/envoy/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:92] StreamSecrets gRPC config stream closed: 2, open deny: no such file or directory
```

Not sure if that is a concern or not

**Comment by ramaraochavali** [comment]
> One thing I noticed is we pretty heavily retry:

May be SDS implementation should do a back-off after error?

**Comment by ramaraochavali** [comment]
@howardjohn  Should we merge this and handle the retry behaviour in a different issue - because I think that issue will come for any error - not just the file error?

**Comment by howardjohn** [comment]
Yes sorry didn't mean to block on that was just a comment. I don't think
I'm an owner for these files

**Comment by ramaraochavali** [comment]
@howardjohn No problem. @istio/wg-security-maintainers PTAL

**Comment by ramaraochavali** [comment]
@istio/wg-security-maintainers gentle ping. Can you PTAL?

**Comment by ramaraochavali** [comment]
@myidpt Yes. I am planning to implement that once this is merged

**Pull request merged**